Yoast/wordpress-seo
# Bug fix 8634: bring back use of `$sort_flags` for `array_unique()`

## Summary

This PR can be summarized in the following changelog entry:
* _Fixed fatal error in sitemap generation for websites using image galleries_

## Relevant technical choices:

Related to #8530 and #8529.

The simple solution would be to just bring back the use of the `$sort_flags` parameter and to ignore PHP 5.2.4 to 5.2.8 which is a minute portion of users anyway.

However, as PHP 5.2.4 - 5.2.8 are still explicitly supported, a more "proper" solution needed to be found.

### Technical explanation of the current proposed solution:

As of PHP 5.2.9, the `array_unique()` function has a second parameter called `$sort_flags`. Using this parameter in PHP < 5.2.9 would result in a `Warning: Wrong parameter count for array_unique()` warning and a `null` result from the function call.

Pre-PHP 5.2.9, under the hood, PHP would use `SORT_STRING` as the sort flag. In PHP 5.2.9 this was changed to `SORT_REGULAR`. In PHP 5.2.10, this was changed back to `SORT_STRING` to undo the backward compatibility break.

The `SORT_STRING` default means that the values are explicitly cast to string and only then compared with each other to determine whether they are equal/duplicates.

The problem with this is that - as of PHP 5.2.0 - casting an object to a string will result in a `Catchable fatal error: Object of class ClassName could not be converted to string`.

Ok, so now we know that, we look at the function itself.

The function retrieves a set of objects through the `get_gallery_attachments()` method and stores them in an array.

So using the default `SORT_STRING` will always result in the catchable fatal error.

Using `SORT_REGULAR` no type casting is executed and the array is semi-undupped - I say semi-undupped as PHP is known for not being able to compare objects well in this manner, but that's for another issue -.

So using `SORT_REGULAR` as the `$sort_flag` when the PHP version allows for it, is the best solution for now.

For PHP < 5.2.9 there is no reliable way to undup the array without generating PHP errors or resorting to a custom array unduplication function. which considering how few users we're talking about is unwarranted, so leaving the array as is, is the best solution, even though it will likely result in duplicate entries in the sitemap.

As for the `PHP_VERSION_ID` constant which I've used: this is only available as of PHP 5.2.7.
However, WordPress back-fills this constant in the `wp-includes/random_compat/random.php` file which is included from the `wp-includes/compat.php` file whenever PHP < 7.0 is detected.
The `compat.php` file is loaded early by WP - before any plugins are loaded -, so the `PHP_VERSION_ID` constant will be available by the time this method is called.

Supporting example code: https://3v4l.org/nFWdQ

Refs:
* http://phpcheatsheets.com/test/string_casting/
* http://php.net/manual/en/function.array-unique.php
* http://php.net/manual/en/reserved.constants.php
* https://core.trac.wordpress.org/browser/trunk/src/wp-includes/compat.php#L442
* https://core.trac.wordpress.org/browser/trunk/src/wp-includes/random_compat/random.php#L32

### Regarding Tide:

The original fix was made because of an error thrown up by Tide.
If/when this fix gets merged, Tide may start throwing an additional error for the usage of `PHP_VERSION_ID`.

I have previously pointed the Tide team to a PHPCompatibility ruleset which takes backfills from WP into account. This should be used by Tide and if the `PHP_VERSION_ID` warning starts getting thrown, it just means they haven't implemented this yet.

As for the error regarding the function parameter, I have added the appropriate PHPCS 3.2.0+ compatible whitelist comment for it.
This will be respected by YoastCS 0.5 once WPSEO starts using YoastCS 0.5.

As for whether Tide will respect this whitelist comment: I can't say.
I would hope so, but I can also easily see problems with this as if they would respect all PHPCS whitelist comments, it would also become extremely easy for plugin/theme authors to "game the system", by adding an `phpcs:ignoreFile` comment at the top of each file, so I'm not sure how they will end up handling this.

N.B.: the aforementioned _PHPCompatibility ruleset which takes backfills from WP into account_ is included in the renewed YoastCS ruleset for YoastCS 0.5.


## Test instructions

This PR can be tested by following these steps using the `trunk` branch:
* Make sure there are image galleries in pages/posts in the test WP setup.
* Clear the sitemap cache
* Open up a sitemap file to see it being empty / displaying the fatal error

Rinse & repeat using this branch.

As a side-note: it might be a good idea to add some unit tests for image galleries to ensure that these kind of issue will be safeguarded against in the future.

Fixes #8634
